File: python/tests/test_cross_language_golden.py
"""Cross-language golden-value tests for v2.2 paths.

Hardcodes outputs of representative v2.2 computations on deterministic
inputs (no RNG). The companion MATLAB file
``matlab/tests/test_cross_language_golden.m`` hardcodes the same
values; running both pins down cross-language numerical agreement to
1e-8 relative on the v2.2 surface (orbit cosine similarity SA + MA
including the safe/unsafe hybrid, Rényi-2 entropy SA + MA, orbit-path
:func:`tensor_harmonicity`, and orbit-path :func:`eval_exp_tens`).

Inputs use ``method='mobius'`` on the cosine cases so the Möbius method's
machinery is genuinely exercised rather than the dispatcher's
cost-model fallback to pairwise. Sigmas are chosen to keep values
well-conditioned (away from FP underflow); a 1e-8 relative tolerance
is the standard used elsewhere in the v22 suite.

To regenerate the golden values (e.g. after a deliberate algorithm
change), run this file as a script: ``python -m
tests.test_cross_language_golden`` prints the freshly computed
values for both languages to mirror.
"""

import numpy as np
import pytest

import mpt
from mpt.tensor import build_exp_tens, cos_sim_exp_tens, eval_exp_tens
from mpt.entropy import entropy_exp_tens
from mpt.harmony import tensor_harmonicity


# Tolerance for cross-language equality. 1e-8 is the standard used
# throughout the v22 suite (orbit-vs-pairwise on shared regimes).
RTOL = 1e-8
ATOL = 1e-12


# ----------------------------------------------------------------------
# Case A: SA cosSim, abs r=3, Möbius method
# ----------------------------------------------------------------------

def test_golden_sa_cossim_abs_r3():
    """Major triad vs minor triad, abs r=3, sigma=80 cents (orbit)."""
    p1 = np.array([0.0, 400.0, 700.0])
    p2 = np.array([0.0, 300.0, 700.0])
    w = np.array([1.0, 1.0, 1.0])
    s = cos_sim_exp_tens(
        p1, w, p2, w, 80.0, 3, False, False, 0.0,
        method='mobius', verbose=False,
    )
    GOLDEN = 0.67614851033133
    assert abs(s - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


# ----------------------------------------------------------------------
# Case B: SA cosSim, rel r=3 periodic, Möbius method
# ----------------------------------------------------------------------

def test_golden_sa_cossim_rel_r3_per():
    """Major triad vs minor triad, rel r=3, period=1200, sigma=80."""
    p1 = np.array([0.0, 400.0, 700.0])
    p2 = np.array([0.0, 300.0, 700.0])
    w = np.array([1.0, 1.0, 1.0])
    s = cos_sim_exp_tens(
        p1, w, p2, w, 80.0, 3, True, True, 1200.0,
        method='mobius', verbose=False,
    )
    GOLDEN = 0.98878587398645
    assert abs(s - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


# ----------------------------------------------------------------------
# Case C: MA cosSim with ragged-K hybrid (the killer cross-language case)
# ----------------------------------------------------------------------

def test_golden_ma_cossim_ragged_k_hybrid():
    """Mixed safe/unsafe ragged-K MA (single attribute, K=8, r=3,
    two unsafe events on each side via NaN padding to K_eff=4).

    Exercises the per-event safe/unsafe partition: safe pairs go
    through the vectorised orbit, unsafe pairs through direct
    enumeration. Identical hybrid logic in both languages.
    """
    P_x = np.array([
        [50.0,  100.0,  200.0,  300.0,  400.0,  500.0],
        [150.0, 250.0,  350.0,  450.0,  550.0,  650.0],
        [350.0, 450.0,  550.0,  650.0,  750.0,  850.0],
        [550.0, 650.0,  750.0,  850.0,  950.0, 1050.0],
        [np.nan, 850.0,  950.0,  np.nan, 1150.0, 1250.0],
        [np.nan, 1050.0, 1150.0, np.nan, 1350.0, 1450.0],
        [np.nan, 1250.0, 1350.0, np.nan, 1550.0, 1650.0],
        [np.nan, 1450.0, 1550.0, np.nan, 1750.0, 1850.0],
    ])  # (8, 6); columns 0 and 3 have K_eff=4 (unsafe at r=3)
    W_x = np.where(np.isnan(P_x), np.nan, 1.0)
    P_y = P_x + 50.0
    W_y = W_x.copy()

    dx = build_exp_tens([P_x], [W_x], [25.0], [3], 
                        [False], [False], [0.0], verbose=False)
    dy = build_exp_tens([P_y], [W_y], [25.0], [3], 
                        [False], [False], [0.0], verbose=False)
    s = cos_sim_exp_tens(dx, dy, method='mobius', verbose=False)
    GOLDEN = 0.12066345091832
    assert abs(s - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


# ----------------------------------------------------------------------
# Case D: SA entropy Rényi-2, abs r=2
# ----------------------------------------------------------------------

def test_golden_sa_renyi2_abs_r2():
    """Rényi-2 entropy of a major-triad density, r=2 abs, sigma=20."""
    p = np.array([0.0, 400.0, 700.0])
    w = np.array([1.0, 1.0, 1.0])
    H = entropy_exp_tens(
        p, w, 20.0, 2, False, False, 0.0,
        method='renyi2', base=2,
    )
    GOLDEN = 14.88031481996820
    assert abs(H - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


# ----------------------------------------------------------------------
# Case E: MA entropy Rényi-2 (Möbius method)
# ----------------------------------------------------------------------

def test_golden_ma_renyi2():
    """MA Rényi-2 entropy on a 2-attribute density (pitch + time),
    r=3 on pitch (K=5, safe), r=1 on time."""
    pitch = np.array([
        [0.0,    200.0,  400.0,  600.0],
        [400.0,  600.0,  700.0,  900.0],
        [700.0,  900.0,  1000.0, 1100.0],
        [1000.0, 1200.0, 1300.0, 1400.0],
        [1100.0, 1300.0, 1500.0, 1700.0],
    ])  # (5, 4)
    time = np.array([[0.0, 0.5, 1.0, 1.5]])  # (1, 4)
    H = entropy_exp_tens(
        [pitch, time], None,
        [12.0, 0.05], [3, 1], 
        [False, False], [True, False], [1200.0, 0.0],
        method='renyi2', base=2,
    )
    GOLDEN = 21.64284222436801
    assert abs(H - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


# ----------------------------------------------------------------------
# Case F: tensor_harmonicity (orbit-rel path; bypasses build_exp_tens)
# ----------------------------------------------------------------------

def test_golden_tensor_harmonicity_orbit():
    """Tensor harmonicity of a major triad with default duplicate=K=3."""
    chord = np.array([0.0, 400.0, 700.0])
    h = tensor_harmonicity(chord, None, 12.0, verbose=False)
    GOLDEN = 0.17358467740231
    assert abs(h - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


# ----------------------------------------------------------------------
# Case G: eval_exp_tens at a query point, orbit-rel
# ----------------------------------------------------------------------

def test_golden_eval_exp_tens_rel():
    """Evaluate a 4-partial harmonic-template tensor at a single query
    (sigma=80 to keep value above FP underflow)."""
    tp = np.array([0.0, 1200.0, 1902.0, 2400.0])
    tw = np.array([1.0, 0.5, 0.333, 0.25])
    X = np.array([[400.0], [700.0]])  # 2 x 1 (r=3 rel -> dim=2)
    v = eval_exp_tens(tp, tw, 80.0, 3, True, False, 1200.0, X,
                      verbose=False)
    GOLDEN = 2.07507623760499e-06
    assert abs(v[0] - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


# ----------------------------------------------------------------------
# Case H: SA Shannon entropy r=2, dim=2 (bin-integration path)
# ----------------------------------------------------------------------
# These cases lock in the bin-integration parity for the discrete
# entropy methods. The bin-integration path was added to Python without
# a parallel MATLAB port for a release window; these goldens catch any
# future drift between the per-axis Phi-difference contractions.

def test_golden_sa_shannon_abs_r2_dim2():
    T = build_exp_tens(
        np.array([100., 200., 300.]), None, 20.0, 2, False, False, 0.0,
        verbose=False,
    )
    H = entropy_exp_tens(
        T, method='shannon',
        n_points_per_dim=40, x_min=50.0, x_max=350.0, verbose=False,
    )
    GOLDEN = 9.383611317877847
    assert abs(H - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


def test_golden_sa_normalized_abs_r2_dim2():
    T = build_exp_tens(
        np.array([100., 200., 300.]), None, 20.0, 2, False, False, 0.0,
        verbose=False,
    )
    H = entropy_exp_tens(
        T, method='normalized',
        n_points_per_dim=40, x_min=50.0, x_max=350.0, verbose=False,
    )
    GOLDEN = 0.8815988444951405
    assert abs(H - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


def test_golden_sa_shannon_periodic_r1():
    T = build_exp_tens(
        np.array([0., 3., 7.]), None, 0.7, 1, False, True, 12.0,
        verbose=False,
    )
    H = entropy_exp_tens(
        T, method='shannon',
        n_points_per_dim=24, verbose=False,
    )
    GOLDEN = 4.093676510565166
    assert abs(H - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


def test_golden_ma_shannon_abs_dim2():
    P2 = np.array([[100., 200., 300.], [200., 250., 100.]])
    W2 = np.array([1., 1., 1.])
    dens = build_exp_tens(
        [P2], [W2], [20.0], [2], [False], [False], [0.0],
        verbose=False,
    )
    H = entropy_exp_tens(
        dens, method='shannon',
        n_points_per_dim=40, x_min=50.0, x_max=350.0, verbose=False,
    )
    GOLDEN = 9.347143263809102
    assert abs(H - GOLDEN) < RTOL * abs(GOLDEN) + ATOL


def test_golden_sa_differential_r1():
    """Adaptive differential entropy. Looser tolerance (1e-4) because
    the adaptive convergence stops at the truncation-sigma-anchored
    tolerance rather than machine precision."""
    T = build_exp_tens(
        np.array([0., 400., 700.]), None, 20.0, 1, False, False, 0.0,
        verbose=False,
    )
    h_hat = entropy_exp_tens(T, method='differential', verbose=False)
    GOLDEN = 7.953986161000217
    # Adaptive convergence tolerance is ~exp(-18) ~ 1.5e-8;
    # allow 1e-5 absolute as a comfortable bound.
    assert abs(h_hat - GOLDEN) < 1e-5


def test_golden_ma_differential_dim2():
    """Adaptive differential entropy on a two-attribute (D==2) absolute
    density. Exercises the D==2 leading-axis cell-block streaming in
    _contract_cell_axes (localContractCellAxes in MATLAB). Narrow spans
    and an explicit truncation_sigmas keep the converged grid small, so
    the value is reproducible at the looser adaptive tolerance."""
    p0 = np.array([[0., 120., 260.]])
    p1 = np.array([[0., 80., 170.]])
    w0 = np.array([[1.0, 0.7, 0.5]])
    w1 = np.array([[1.0, 1.0, 1.0]])
    T = build_exp_tens(
        [p0, p1], [w0, w1], [45.0, 35.0], [1, 1],
        [False, False], [False, False], [0.0, 0.0], verbose=False,
    )
    h_hat = entropy_exp_tens(
        T, method='differential', base=2.0,
        truncation_sigmas=3.0, verbose=False,
    )
    GOLDEN = 16.07500552585256
    assert abs(h_hat - GOLDEN) < 1e-4


if __name__ == '__main__':
    import sys
    print("=" * 60)
    print("Generating golden values (Python side)")
    print("=" * 60)
    for name in [n for n in sorted(dir()) if n.startswith('test_golden_')]:
        try:
            globals()[name]()
            print(f"  PASS  {name}")
        except AssertionError as e:
            print(f"  FAIL  {name}: {e}", file=sys.stderr)

File: matlab/tests/test_cross_language_golden.m
%% test_cross_language_golden.m — v3 cross-language equivalence
%
%  Hardcodes outputs of representative v3 computations on
%  deterministic inputs (no RNG). The companion Python file
%  python/tests/test_cross_language_golden.py hardcodes the same
%  values; running both pins down cross-language numerical agreement
%  to 1e-8 relative on the v3 surface (Möbius cosine similarity single-multiset
%  + MA Rényi-2 entropy single-multiset + MA,
%  Möbius-method tensorHarmonicity, and Möbius-method evalExpTens).
%
%  Inputs use 'method', 'mobius' on the cosine cases so the Möbius method
%  Möbius machinery is genuinely exercised rather than the
%  dispatcher's cost-model fallback to Bulger. Sigmas are chosen
%  to keep values well-conditioned (away from FP underflow); a 1e-8
%  relative tolerance is the standard used elsewhere in the v22 suite.
%
%  When regenerating the golden table after a deliberate algorithm
%  change, update the values in BOTH this file and the Python mirror
%  to the same number of digits.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    % Defaults isolation when run standalone (when invoked from
    % test_mpt.m the outer wrapper has already isolated defaults).
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

RTOL = 1e-8;
ATOL = 1e-12;

%% ---- Case A: single-multiset cosSim, abs r=3, Möbius ----

p1 = [0; 400; 700];
p2 = [0; 300; 700];
w  = [1; 1; 1];
sA = cosSimExpTens(p1, w, p2, w, 80, 3, false, false, 0, ...
    'method', 'mobius', 'verbose', false);
GOLDEN_A = 0.67614851033133;
results{end+1, 1} = 'cross-language golden A: single-multiset cosSim abs r=3 Möbius';
results{end, 2}   = abs(sA - GOLDEN_A) < RTOL * abs(GOLDEN_A) + ATOL;

%% ---- Case B: single-multiset cosSim, rel r=3 per, Möbius ----

sB = cosSimExpTens(p1, w, p2, w, 80, 3, true, true, 1200, ...
    'method', 'mobius', 'verbose', false);
GOLDEN_B = 0.98878587398645;
results{end+1, 1} = 'cross-language golden B: single-multiset cosSim rel r=3 per Möbius';
results{end, 2}   = abs(sB - GOLDEN_B) < RTOL * abs(GOLDEN_B) + ATOL;

%% ---- Case C: MA cosSim ragged-K ----

P_x = [ 50  100  200  300  400  500;
       150  250  350  450  550  650;
       350  450  550  650  750  850;
       550  650  750  850  950 1050;
       NaN  850  950  NaN 1150 1250;
       NaN 1050 1150  NaN 1350 1450;
       NaN 1250 1350  NaN 1550 1650;
       NaN 1450 1550  NaN 1750 1850];   % (8, 6)
W_x = ones(size(P_x));
W_x(isnan(P_x)) = NaN;
P_y = P_x + 50;
W_y = W_x;
dx = buildExpTens({P_x}, {W_x}, 25, 3, false, false, 0, ...
    'verbose', false);
dy = buildExpTens({P_y}, {W_y}, 25, 3, false, false, 0, ...
    'verbose', false);
sC = cosSimExpTens(dx, dy, 'method', 'mobius', 'verbose', false);
GOLDEN_C = 0.12066345091832;
results{end+1, 1} = 'cross-language golden C: MA cosSim ragged-K';
results{end, 2}   = abs(sC - GOLDEN_C) < RTOL * abs(GOLDEN_C) + ATOL;

%% ---- Case D: single-multiset entropy Rényi-2, abs r=2 ----

HD = entropyExpTens(p1, w, 20, 2, false, false, 0, ...
    'method', 'renyi2', 'base', 2);
GOLDEN_D = 14.88031481996820;
results{end+1, 1} = 'cross-language golden D: single-multiset entropy Rényi-2 abs r=2';
results{end, 2}   = abs(HD - GOLDEN_D) < RTOL * abs(GOLDEN_D) + ATOL;

%% ---- Case E: MA entropy Rényi-2 ----

pitch = [   0   200   400   600;
          400   600   700   900;
          700   900  1000  1100;
         1000  1200  1300  1400;
         1100  1300  1500  1700];   % (5, 4)
time = [0 0.5 1.0 1.5];              % (1, 4)
HE = entropyExpTens({pitch, time}, [], ...
    [12, 0.05], [3, 1], ...
    [false, false], [true, false], [1200, 0], ...
    'method', 'renyi2', 'base', 2);
GOLDEN_E = 21.64284222436801;
results{end+1, 1} = 'cross-language golden E: MA entropy Rényi-2';
results{end, 2}   = abs(HE - GOLDEN_E) < RTOL * abs(GOLDEN_E) + ATOL;

%% ---- Case F: tensorHarmonicity Möbius method ----

hF = tensorHarmonicity([0; 400; 700], [], 12, 'verbose', false);
GOLDEN_F = 0.17358467740231;
results{end+1, 1} = 'cross-language golden F: tensorHarmonicity Möbius';
results{end, 2}   = abs(hF - GOLDEN_F) < RTOL * abs(GOLDEN_F) + ATOL;

%% ---- Case G: evalExpTens at single query, rel Möbius ----

tp = [0; 1200; 1902; 2400];
tw = [1; 0.5; 0.333; 0.25];
X  = [400; 700];   % 2 x 1 (r=3 rel -> dim=2)
v = evalExpTens(tp, tw, 80, 3, true, false, 1200, X, 'verbose', false);
GOLDEN_G = 2.07507623760499e-06;
results{end+1, 1} = 'cross-language golden G: evalExpTens rel orbit';
results{end, 2}   = abs(v - GOLDEN_G) < RTOL * abs(GOLDEN_G) + ATOL;

%% ---- Case H: single-multiset Shannon entropy abs r=2 dim=2 (bin-integration path) ----
% These cases lock in the bin-integration parity for the discrete entropy
% methods. The bin-integration path was added to Python without a
% parallel MATLAB port for a release window; these goldens catch any
% future drift between the per-axis Phi-difference contractions.

dens_h = buildExpTens([100; 200; 300], [], 20, 2, false, false, 0, ...
    'verbose', false);
HH = entropyExpTens(dens_h, 'method', 'shannon', ...
    'nPointsPerDim', 40, 'xMin', 50, 'xMax', 350, 'verbose', false);
GOLDEN_H = 9.383611317877847;
results{end+1, 1} = 'cross-language golden H: single-multiset shannon abs r=2 dim=2 bin-integration';
results{end, 2}   = abs(HH - GOLDEN_H) < RTOL * abs(GOLDEN_H) + ATOL;

%% ---- Case I: single-multiset normalized (Pielou ratio) ----

HI = entropyExpTens(dens_h, 'method', 'normalized', ...
    'nPointsPerDim', 40, 'xMin', 50, 'xMax', 350, 'verbose', false);
GOLDEN_I = 0.8815988444951405;
results{end+1, 1} = 'cross-language golden I: single-multiset normalized abs r=2 dim=2';
results{end, 2}   = abs(HI - GOLDEN_I) < RTOL * abs(GOLDEN_I) + ATOL;

%% ---- Case J: single-multiset Shannon periodic r=1 ----

dens_j = buildExpTens([0; 3; 7], [], 0.7, 1, false, true, 12, ...
    'verbose', false);
HJ = entropyExpTens(dens_j, 'method', 'shannon', ...
    'nPointsPerDim', 24, 'verbose', false);
GOLDEN_J = 4.093676510565166;
results{end+1, 1} = 'cross-language golden J: single-multiset shannon periodic r=1';
results{end, 2}   = abs(HJ - GOLDEN_J) < RTOL * abs(GOLDEN_J) + ATOL;

%% ---- Case K: MA Shannon abs dim=2 ----

P2 = [100, 200, 300; 200, 250, 100];
W2 = [1, 1, 1];
dens_k = buildExpTens({P2}, {W2}, 20, 2, false, false, 0, ...
    'verbose', false);
HK = entropyExpTens(dens_k, 'method', 'shannon', ...
    'nPointsPerDim', 40, 'xMin', 50, 'xMax', 350, 'verbose', false);
GOLDEN_K = 9.347143263809102;
results{end+1, 1} = 'cross-language golden K: MA shannon abs dim=2';
results{end, 2}   = abs(HK - GOLDEN_K) < RTOL * abs(GOLDEN_K) + ATOL;

%% ---- Case L: single-multiset differential entropy (adaptive) ----
% Adaptive convergence tolerance is ~exp(-18) ~ 1.5e-8;
% allow 1e-5 absolute as a comfortable bound.

dens_l = buildExpTens([0; 400; 700], [], 20, 1, false, false, 0, ...
    'verbose', false);
HL = entropyExpTens(dens_l, 'method', 'differential', 'verbose', false);
GOLDEN_L = 7.953986161000217;
results{end+1, 1} = 'cross-language golden L: single-multiset differential r=1 adaptive';
results{end, 2}   = abs(HL - GOLDEN_L) < 1e-5;

%% ---- Case M: MA differential entropy (adaptive, D==2) ----
% Two attributes, r=1 each (D==2). Exercises the D==2 leading-axis
% cell-block streaming in localContractCellAxes. Narrow spans and an
% explicit truncationSigmas keep the converged grid small, so the value
% is reproducible at the looser adaptive tolerance.

P0_m = [0, 120, 260];
P1_m = [0, 80, 170];
W0_m = [1.0, 0.7, 0.5];
W1_m = [1.0, 1.0, 1.0];
HM = entropyExpTens({P0_m, P1_m}, {W0_m, W1_m}, [45, 35], [1, 1], ...
    [false, false], [false, false], [0, 0], ...
    'method', 'differential', 'base', 2, ...
    'truncationSigmas', 3.0, 'verbose', false);
GOLDEN_M = 16.07500552585256;
results{end+1, 1} = 'cross-language golden M: MA differential D==2 adaptive';
results{end, 2}   = abs(HM - GOLDEN_M) < 1e-4;

%% ---- Standalone summary ----

if standalone
    nPass = sum(cellfun(@(x) isequal(x, true), results(:, 2)));
    nFail = numel(results(:, 1)) - nPass;
    fprintf('\n=== test_cross_language_golden: %d passed, %d failed (of %d) ===\n', ...
        nPass, nFail, numel(results(:, 1)));
    % Restore caller's pre-test defaults eagerly
    % (fires the helper's onCleanup destructor before
    % the conditional error below halts execution).
    clear cleanupDefaults
    if nFail > 0
        for ii = 1:size(results, 1)
            if ~isequal(results{ii, 2}, true)
                fprintf('  FAIL  %s\n', results{ii, 1});
            end
        end
    end
end
